File: apps/cargo-tracker/app/cargotracker/tracking/domain/model/repositories/TrackingActivityRepository.scala
package cargotracker.tracking.domain.model.repositories

import cargotracker.tracking.domain.model.aggregates.TrackingActivity
import cargotracker.tracking.domain.model.valueobjects.{TrackingBookingId, TrackingNumber}

/** Tracking Context のリポジトリポート（ヘキサゴナル）。
  *
  * 採番は `nextTrackingNumber` で行う（実装は `BIGSERIAL` を `TN-NNNNNN` に整形）。
  */
trait TrackingActivityRepository:
  def nextTrackingNumber(): TrackingNumber
  def findByTrackingNumber(trackingNumber: TrackingNumber): Option[TrackingActivity]
  def findByBookingId(bookingId: TrackingBookingId): Option[TrackingActivity]

  /** 集約ルートの永続化。新規 INSERT または `version` の差分更新のみを行う（events は appendEvent 経由）。 */
  def save(activity: TrackingActivity): Unit

  /** 集約配下の新規イベント 1 件を追記し、`transportStatus` と `version` を同時更新する（楽観ロック）。
    *
    * 戻り値: 新しい version を持つ TrackingActivity。呼出側はこれを引数の `activity` の代わりに後続処理で再利用すること（H1 対応）。
    */
  def appendEvent(
      activity: TrackingActivity,
      newEvent: cargotracker.tracking.domain.model.entities.TrackingActivityEvent
  ): TrackingActivity

  /** 例外イベントを追記する（IT7 US19/US20）。`version` を楽観ロックで更新する。 */
  def appendException(
      activity: TrackingActivity,
      newException: cargotracker.tracking.domain.model.entities.TrackingExceptionEvent
  ): TrackingActivity

  /** 指定 index の例外を解決済に更新する（IT7 US19/US20）。 */
  def updateExceptionResolution(
      activity: TrackingActivity,
      index: Int,
      resolvedAt: java.time.Instant,
      resolutionNotes: String
  ): TrackingActivity

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/tracking/domain/repository/TrackingActivityRepository.java
package com.example.cargotracker.tracking.domain.repository;

import com.example.cargotracker.tracking.domain.model.TrackingActivity;
import com.example.cargotracker.tracking.domain.model.TrackingBookingId;
import com.example.cargotracker.tracking.domain.model.TrackingNumber;
import java.util.Optional;
import javax.annotation.CheckReturnValue;

/** 追跡レコードの出力ポート。実装はインフラ層に置く（DIP）。 */
public interface TrackingActivityRepository {

    /** 追跡を始める（US14）。 */
    void save(TrackingActivity activity);

    /**
     * 輸送状態とイベントを保存する（US15）。
     *
     * <p><strong>状態とイベントを 1 つの操作として書く。</strong> 片方だけが残ると、
     * 「積込済なのにイベントが無い」「イベントはあるが未受取のまま」になる。
     *
     * @return 他の更新が先行していれば {@code false}（楽観的ロック）
     */
    @CheckReturnValue
    boolean update(TrackingActivity activity);

    Optional<TrackingActivity> findByTrackingNumber(TrackingNumber trackingNumber);

    Optional<TrackingActivity> findByBookingId(TrackingBookingId bookingId);

    /**
     * 例外を持つ追跡番号（IT13 レビュー C4）。
     *
     * <p><strong>1 件ずつ聞かない。</strong> 一覧の行数だけ問い合わせが飛ぶ。
     */
    java.util.Set<String> findTrackingNumbersWithUnresolvedException(
            java.util.Collection<String> trackingNumbers);
}
